bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/VerisureSession.java

```diff
@@ -0,0 +1,906 @@
+        return csrf;
+    }
+
+    public @Nullable String getPinCode() {
+        return pinCode;
+    }
+
+    public String getApiServerInUse() {
+        return apiServerInUse;
+    }
+
+    public void setApiServerInUse(String apiServerInUse) {
+        this.apiServerInUse = apiServerInUse;
+    }
+
+    public String getNextApiServer() {
+        apiServerInUseIndex++;
+        if (apiServerInUseIndex > (APISERVERLIST.size() - 1)) {
+            apiServerInUseIndex = 0;
+        }
+        return APISERVERLIST.get(apiServerInUseIndex);
+    }
+
+    public void configureInstallationInstance(BigDecimal installationId) {
+        logger.debug("Attempting to fetch CSRF and configure installation instance");
+        try {
+            csrf = getCsrfToken(installationId);
+            logger.debug("Got CSRF: {}", csrf);
+            // Set installation
+            String url = SET_INSTALLATION + installationId.toString();
+            logger.debug("Set installation URL: {}", url);
+            httpClient.GET(url);
+        } catch (ExecutionException | InterruptedException | TimeoutException e) {
+            logger.warn("Caught an Exception {}", e.getMessage(), e);
+        }
+    }
+
+    public @Nullable String getCsrfToken(BigDecimal installationId) {
+        String html = null;
+        String url = SETTINGS + installationId.toString();
+        logger.debug("Settings URL: {}", url);
+
+        try {
+            ContentResponse resp = httpClient.GET(url);
+            html = resp.getContentAsString();
+            logger.trace("{} html: {}", url, html);
+        } catch (ExecutionException | InterruptedException | TimeoutException e) {
+            logger.warn("Caught an Exception {}", e.getMessage(), e);
+            return null;
+        }
+
+        Document htmlDocument = Jsoup.parse(html);
+        Element nameInput = htmlDocument.select("input[name=_csrf]").first();
+        return nameInput.attr("value");
+    }
+
+    public @Nullable String getPinCode(BigDecimal installationId) {
+        return verisureInstallations.get(installationId).getPinCode();
+    }
+
+    private void setPasswordFromCookie() {
+        CookieStore c = httpClient.getCookieStore();
+        List<HttpCookie> cookies = c.get(URI.create("http://verisure.com"));
+        cookies.forEach(cookie -> {
+            logger.debug("Response Cookie: {}", cookie);
+            if (cookie.getName().equals(passwordName)) {
+                password = cookie.getValue();
+                logger.debug("Fetching vid {} from cookie", password);
+            }
+        });
+    }
+
+    private void logTraceWithPattern(int responseStatus, String content) {
+        if (logger.isTraceEnabled()) {
+            String pattern = "(?m)^\\s*\\r?\\n|\\r?\\n\\s*(?!.*\\r?\\n)";
+            String replacement = "";
+            logger.trace("HTTP Response ({}) Body:{}", responseStatus, content.replaceAll(pattern, replacement));
+        }
+    }
+
+    private int areWeLoggedIn() {
```

Comment:
True and fixed!

Reply:
I got a question on usage of DeviceStatusListener in VerisureSession e.g. if this list creation:

```
private final List<DeviceStatusListener> deviceStatusListeners = new CopyOnWriteArrayList<>();
```

should be written like this:

```
private final List<DeviceStatusListener<VerisureThing>> deviceStatusListeners = new CopyOnWriteArrayList<>();
```

Otherwise I get this compiler warning:

```
DeviceStatusListener is a raw type. References to generic type DeviceStatusListener<T> should be 
 parameterized
```

Reply:
Well we were going to have to address that eventually but yes you are correct, it should be written as:
```
private final List<DeviceStatusListener<VerisureThing>> deviceStatusListeners = new CopyOnWriteArrayList<>();
```

Reply:
What about the register/unregister methods in VerisureSession?

```
public boolean unregisterDeviceStatusListener(DeviceStatusListener deviceStatusListener) {
        logger.debug("unregisterDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.remove(deviceStatusListener);
}
```

If a change them accordingly I get a compile error in VTH dispose() method:

```
public boolean unregisterDeviceStatusListener(DeviceStatusListener deviceStatusListener) {
        logger.debug("unregisterDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.remove(deviceStatusListener);
}
```

```
The method unregisterDeviceStatusListener(DeviceStatusListener<VerisureThing>) in the type VerisureSession is not applicable for the arguments 
 (VerisureThingHandler<T>)
```

Reply:
Ugh, this is where generics start to get very annoying. You would need to add generic types to the parameters for methods. We could possibly use wildcards but wildcards can cause a lot of pain I would like you to avoid.
```
<T extends VerisureThing> void registerDeviceStatusListener(DeviceStatusListener<T>)
<T extends VerisureThing> void unregisterDeviceStatusListener(DeviceStatusListener<T>)
```

Reply:
In fact, even that might not work either. So instead just do this:
```
public boolean unregisterDeviceStatusListener(DeviceStatusListener<? extends VerisureThing> deviceStatusListener) {
        logger.debug("unregisterDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.remove((DeviceStatusListener<VerisureThing>) deviceStatusListener);
}
```
Lets just cast the DeviceStatusListener type before inserting or removing from `deviceStatusListeners`. Then just suppress the remaining warnings because I don't know of any better way to deal with this.

Reply:
If you are lucky it might let you do it without casting.

Reply:
What do you think about this construction for the unregister/register methods in VersureSession:

```
public <T extends VerisureThing> boolean unregisterDeviceStatusListener(
            DeviceStatusListener<T> deviceStatusListener) {
        logger.debug("unregisterDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.remove(deviceStatusListener);
}

public <T extends VerisureThing> boolean registerDeviceStatusListener(
            DeviceStatusListener<T> deviceStatusListener) {
        logger.debug("registerDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.add(deviceStatusListener);
 }
```

If that is OK, the last thing that I struggle with is the list definition:

```
private final List<DeviceStatusListener> deviceStatusListeners = new CopyOnWriteArrayList<>();
```

Using this definition it works, however shouldn't that also contain some kind of generics definition?

If I use only `private final List<DeviceStatusListener<VerisureThing>>`I get a compile error on:

```
public <T extends VerisureThing> boolean registerDeviceStatusListener(
            DeviceStatusListener<T> deviceStatusListener) {
        logger.debug("registerDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.add(deviceStatusListener);
}
```

```
The method add(DeviceStatusListener<VerisureThing>) in the type List<DeviceStatusListener<VerisureThing>> is not applicable for the arguments 
 (DeviceStatusListener<T>)
```

But why does not this compile:

```
private <T extends VerisureThing> final List<DeviceStatusListener<T>> deviceStatusListeners = new CopyOnWriteArrayList<>();
```

Compile error:

```
Multiple markers at this line
	- T cannot be resolved to a type
	- Syntax error on tokens, ReferenceType expected 
	 instead
```

Reply:
I will push my latest updates so that you more easily can see the latest changes made.

Reply:
What if you do my other suggestion?
```
public boolean unregisterDeviceStatusListener(DeviceStatusListener<? extends VerisureThing> deviceStatusListener) {
        logger.debug("unregisterDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.remove((DeviceStatusListener<VerisureThing>) deviceStatusListener);
}
```

Reply:
```
private <T extends VerisureThing> final List<DeviceStatusListener<T>> deviceStatusListeners = new CopyOnWriteArrayList<>();
```
Doesn't compile because you can't use `<T extends VerisureThing>` on fields.
What you might be trying to do is:
```
private final List<DeviceStatusListener<? extends VerisureThing>> deviceStatusListeners = new CopyOnWriteArrayList<>();
```
But the problem with the defining will wildcards like that is that the list essentially becomes read only because of the wildcard.
In order to be able to both read and write to that list it would need to be defined without the wildcard:
```
private final List<DeviceStatusListener<VerisureThing>> deviceStatusListeners = new CopyOnWriteArrayList<>();
```

These kinda of restrictions are why I was saying that using wildcards would make things painful 😁 .

Reply:
Then I then can have the list definition as:

`private final List<DeviceStatusListener<VerisureThing>> deviceStatusListeners = new CopyOnWriteArrayList<>();`

But I get a compile warning on `deviceStatusListeners.add`:

```
public boolean registerDeviceStatusListener(DeviceStatusListener<? extends VerisureThing> deviceStatusListener) {
        logger.debug("registerDeviceStatusListener for listener {}", deviceStatusListener);
        return deviceStatusListeners.add((DeviceStatusListener<VerisureThing>) deviceStatusListener);
}
```

```
	- Type safety: Unchecked cast from DeviceStatusListener<capture#4-of ? extends VerisureThing> to 
	 DeviceStatusListener<VerisureThing>
```

Then we should use your suggestion, since that only yields a warning but the List definition uses Genererics, but why is `<? extends VerisureThing>` preferable compared to `<T extends VerisureThing>` ?

? is some kind of wildcard definition as I understand it that states any type that extends VerisureThing.

Reply:
Well you use `<T extends VerisureThing>` when you plan on declaring `T` variables or thing like `Class<T>` in the method whereas you use `<? extends VerisureThing>` when you don't care what type it is because your method won't be dealing with the type directly.